### ISSUE TITLE
Excel-DNA Roadmap Discussion

### ISSUE THREAD

**Issue by govert** [Excel-DNA/ExcelDna#189]
How should Excel-DNA evolve over the coming years?

I can imagine the following versions, with some features that may be progressively realised over the coming years.

### version 0.35

- [x]  Improve build tasks - more reliable clean-up and debugger detection.
- [x]  Improve RTD and async QueueAsMacro reliability
- [x]  Improve install process of ExcelDna.AddIn NuGet package (now requires NuGet 2.5)
- [x]  Change how ExcelDnaUtil.Application works in Protected View - try harder but don't cache
- [x]  Call UnhandledExceptionHandler for macros (ExcelCommands) too

Released as soon as possible - end of May / early June 2018.

### version 0.36

* Increase IntelliSense flexibility
* Improved unloading support
* Improve build process (Excel-DNA itself, not user add-ins)

Some example issues:

- [  ] #85 Store and return untruncated attribute descriptions
- [  ] #67 Allow complete unloading of add-in - implement DllCanUnloadNow()
- [  ] #49 Fix AppVeyor build

### version 1.0

Final version to support .NET versions < 4.0 and Excel versions < 2007

Maybe by 2019?

### version 1.1

First version with only .NET 4.0+ and Excel 2007+, preserving interface (existing add-ins should work / easily recompile)

* Rework marshaling implementation
* Add support for high-performance marshaling with Span<T>, in-place parameters etc.
* RTD marshaling improvements (prevent multiple result copies).

### version 2.0

Add support for .NET Core, including pre-compiled add-ins with CoreRT (also allows wabasm support for running as add-ins on other platforms).

Add-In projects might look very different, but we should try to preserve behaviour of the code as far as possible.

Maybe in the early 2020s?

**Comment by caioproiete** [comment]
Love it!

## version 0.36
* Improved unloading support
I'd say this is very high priority and would open the door for a better story at updating Excel-DNA powered add-ins without having to restart Excel, which would be a huge win!

I think if we get this working and are confident enough, it's so important that would be worth releasing it alone as v0.36 and leave whatever else is left for a v0.37 release.


## Possible suggestions for future versions, from v0.37 to v9000 😃 

Some of these will never get done, others might have a better fit as a separate project instead of core Excel-DNA, etc. but for discussion purposes below is some of possible suggestions (and dreams) - the themes and items are not sorted in any particular way... Just a brain dump:

* **Improve the experience of professional developers / teams**
  * Change the lifecycle of add-ins so that we have a single and consistent entry-point (`IExcelAddIn`) and objects get instantiated in the right order - so that when an `ExcelRibbon` gets instantiated, our entry point already ran and set up everything the ribbon needs
  * Add support for Dependency Injection and Unit Testing and make them first-class citizens
  * Add support (and samples) for add-ins using MV\*\*\* patterns
  * Add support for `async` and `await` (e.g. `IExcelDnaAsync` with an `AutoOpen` that returns a `Task`)
  * Improve the API for [ExcelDna.Registration](https://github.com/Excel-DNA/Registration) and make it the default & preferred way for registering functions from File->New Project
  * Add support for mocking Excel as a whole ( 🦄one can dream, right? )

* **Improve the add-in developer experience (all levels)**
  * Add reliable support for add-in deactivation when Excel is being closed, with enough time to do clean-up, free resources, etc.
  * Add TlbExp option to build targets
  * Convert ExcelDna pack into a build task
  * Create Roslyn analyzers to detect common issues in the project (e.g. Your class must be public for the macros to be registered with Excel)
  * Add support for creating Ribbons directly via code, without having to write any XML

* **Keep up with changes in the .NET ecosystem**
  * Add support for the NuGet `PackageReference` format
  * Add support for .NET Standard
  * Add support for developing add-ins with Visual Studio Code

* **Improve the add-in end-user experience**
  * Allow users not to have to think about launching the right add-in (32-bit/64-bit) and just do the right thing for them
  * Add support for (automatically) embedding metadata in Excel files, so that when a user sends a spreadsheet to someone that does **not** have the add-in installed, the user (at least) gets a nice message saying that they need to install add-in X

* **New functionality out-of-the-box**
* Implement a `RemoteFunctions` feature similar to [XLLoop](https://github.com/poidasmith/xlloop) providing a spec/protocol so that devs can develop functions in whatever language they want (in the back-end)
* Implement an inter-process communication feature so that different add-ins can talk to each other and dynamically share functionality

**Comment by govert** [comment]
Thanks @caioproiete.

Let me try to put your list into a few buckets:

* **1.0** - a good feature to work on now, hopefully targeting 1.0 time frame ('final' version with current structure)
* **1.1** - a good feature to work on now, hopefully targeting 1.1 time frame (mostly compatible with the current project, extra features and 
* **2.0** - something that might targeting a 2.0 time frame, where we might consider large changes to the project and add-in structure
* **Sample / Independent Project** - means an orthogonal feature that does not need changes to the core project, and can be done based on the current project (i.e. based on the version 1.0 feature set)
* **Not sure** - means I'm not yet convinced that this is a good idea, and we should discuss before any work is done.

> 
> ---
> 
> * **Improve the experience of professional developers / teams**
>   * Change the lifecycle of add-ins so that we have a single and consistent entry-point (`IExcelAddIn`) and objects get instantiated in the right order - so that when an `ExcelRibbon` gets instantiated, our entry point already ran and set up everything the ribbon needs to work

**Not sure** / **2.0** - I still don't understand what you need here. We can restart a separate discussion for this.

>   * Add support for Dependency Injection and Unit Testing and make them first-class citizens

**Not sure** / **2.0** - I'm open to some suggestions for improvement that will make this easier to bolt on externally, but definitely against any built-in support for DI.

>   * Add support (and samples) for add-ins using MV\*\*\* patterns

**Not sure** / **2.0** - I'm open to some suggestions for improvement that will make this easier to bolt on externally, but definitely against any built-in support for these patterns.

>   * Add support for `async` and `await` (e.g. `IExcelDnaAsync` with an `AutoOpen` that returns a `Task`)

**Not sure** - I don't understand what is needed here.

>   * Add support for App.config XTD transformations out-of-the-box

**Not sure** - I don't know what this means, or what parts of Excel-DNA it would impact. Open to discussion.

>   * Improve the API for [ExcelDna.Registration](https://github.com/Excel-DNA/Registration) and make it the default & preferred way for registering functions from File->New Project

**1.1** - I'm happy with this if we can make it compatible. It requires the .NET 4+ move envisioned as version 1.1.

>   * Add support for mocking Excel as a whole ( 🦄one can dream, right? )

**Sample / Independent Project** - probably you'd split this into a new 'host' that can load and deal with any native .xll add-in (not too hard), and something that mocks the Excel object model (not hard but lots of work - maybe ClosedXML is a start as it already has a C# implementation of much of the object model.

> * **Improve the add-in developer experience (all levels)**
>   * Add reliable support for add-in deactivation when Excel is being closed, with enough time to do clean-up, free resources, etc.

**Sample / 1.0** - Is the notification you get when a COM add-in (e.g. ribbon) is disconnected good enough?

>   * Add TlbExp option to build targets

**1.0** - I'd like this 

>   * Convert ExcelDna pack into a build task

**1.0** - I'd like this. Should keep separate ExcelDnaPack as an .exe too (it might reference and load the build task library with the 'real' implementation. ** Maybe add code signing  to this too? **

>   * Create Roslyn analyzers to detect common issues in the project (e.g. Your class must be public for the macros to be registered with Excel)

**Not sure** - I don't know what this means, or what parts of Excel-DNA it would impact. Open to discussion.

>   * Add support for creating Ribbons directly via code, without having to write any XML

**Sample / Independent Project** - Nothing stops this now. Even a visual designer might be nice to incorporate.

> 
> ---
> 
> * **Improve the add-in end-user experience**
>   * Allow users not to have to think about launching the right add-in (32-bit/64-bit) and just do the right thing for them

**Not sure / 2.0** - I don't have good technical plans for this. We can investigate again whether the magic DllExport feature of .NET will allow the right runtime to be picked automatically, but I doubt it. An alternative is to load a COM add-in that then does RegisterXLL, but the overall user experience is even worse. Another alternative is to make a small start-up .exe which asks the user which Excel to load stuff into, or starts a new Excel and loads the right add-in. No support in Excel-DNA needed for the last two suggestions.

>   * Resolve the issue(s) that causes add-ins to get disabled sometimes and exceptions are thrown in the user's face

**1.0 / 1.1** - I'd like this. One options is to wrap all ribbon calls ourselves by implementing an IDispatch wrapper. That allows us to catch any unhandled exceptions.

>   * Add support for (automatically) embedding metadata (and code) in Excel files, so that when a user sends a spreadsheet to someone that does **not** have the add-in installed, the user (at least) gets a nice message saying that they need to install add-in X for the spreadsheet to work

**Sample / Independent Project** - I like this. Nothing stops this now as a little sample. 

> * **Keep up with changes in the .NET ecosystem**
>   * Add support for the NuGet `PackageReference` format

**1.1/2.0** - I'm not sure of the short-term benefits. Depends on the technical issues involved.

>   * Add support for .NET Standard

**1.0 / 1.1 / Not sure** - AFAIK we can already consume .NET Standard libraries. I don't know if there is some issue with assembly redirection - we should check this. I'm not sure what the implication or value is of making ExcelDna.Integration target .NET Standard. 

>   * Add support for developing add-ins with Visual Studio Code

**Not Sure / Sample / Independent Project** - What does this mean?

> 
> ---
> 
> * **New functionality out-of-the-box**
>   * Implement a `RemoteFunction` feature similar to [XLLoop](https://github.com/poidasmith/xlloop) providing a spec/protocol so that devs can develop functions in whatever language they want (in the back-end)

**Sample / Independent Project** - I like this, but not as part of the core project. Nothing stops this now and I've seen a few implementation with Excel-DNA. 

>   * Implement an inter-process communication feature so that different add-ins can talk to each other and dynamically share functionality

**Not Sure** - I'm not sure I'd encourage this.

> 
> ---
> 
> * **Community / Open-source / Support**
>   * [Finish the conversion of the website to GitHub Pages](https://github.com/Excel-DNA/ExcelDna/issues/129) and launch it. Let our developers to update the website themselves and send pull-requests

**Not Sure** - My hesitation here has to do with the split between what is 'mine' and what is the 'open-source' project. I need a little bit more time to think about this.

>   * Implement a better documentation / wiki integrated with the GitHub Pages site above using [DocFx](https://dotnet.github.io/docfx/) and let users update the documentation themselves in-place directly on the GitHub web user interface, [just like Microsoft is doing now](https://docs.microsoft.com/en-us/dotnet/framework/):
>      [![image](https://user-images.githubusercontent.com/177608/40218360-3da4e9a8-5a47-11e8-84af-143216fd01c5.png)](https://docs.microsoft.com/en-us/dotnet/framework/)

**Yes** - I'd like this. I think we should start a documentation site that is separate from the GitHub Pages site (e.g. at docs.excel-dna.net).

>   * Wholesale copy contents of valuable answers to common problems and/or guidance/recommendations ([this](https://groups.google.com/forum/#!topic/exceldna/WCCvYvqqhUU) for example) to the wiki implemented above to stop answering the same questions over and over. Aim to always answer questions with a link to a wiki article that either existed or was just created for the purpose of the response. "[We have a finite number of keystrokes left in our hands before we die](http://keysleft.com/)"

**Yes** - I agree.

>   * Retire the [Excel-DNA Google Group](https://groups.google.com/group/exceldna) and make [StackOverflow](https://stackoverflow.com/questions/tagged/excel-dna) the official channel for asking questions and getting support. SO is much better ranked on search engines making it easier to find answers to problems and decreases repeated questions being asked over and over like in the Goggle Group today

**No way** - I won't consider this at the moment, even though I answer questions on both. The Google group is much friendlier and usable than StackOverflow for this kind of thing..

>   * Increase the \# of contributions/pull-requests from developers by leveraging more of GitHub's features
>     * Create issues for everything that gets added to the roadmap so that people can chip in
>     * Allow users to assign issues to themselves, so 
>     * Use "Projects" to better visualize the roadmap on a board as well as a kanban board to better communicate current progress, what needs scoping, what's ready to start (by anyone), etc.
>     * Create tag existing issues as [firsttimersonly](https://www.firsttimersonly.com/), [awesome-for-beginners](https://github.com/MunGell/awesome-for-beginners), [up-for-grabs](https://up-for-grabs.net/), etc. to give our issues more visibility to the dev community and attract newcomers to contribute

**Yes** - I agree and I'm trying.

>   * Promote the use of [Gitter](https://gitter.im/) or [Slack](https://slack.com/) channels for exchanging ideas of possible features, bug fixes, etc. in a more informal way and (sometimes) less asynchronous than GitHub issues

**Not sure** - I prefer asynchronous email communication, and especially the searchable Google group list. I feel like important discussions just get lost in chat rooms.

>   * Ship pre-release versions of the NuGet packages on every pull-request that gets merged, automatically. e.g. Have the AppVeyor publish a pre-release package to a NuGet feed somewhere (e.g. via free/public [MyGet] automatically on every successful build. Let users try new features/fixes as soon as they are merged

**0.36** - I'd like this - it falls under the "Improve build process (Excel-DNA itself, not user add-ins)" item I've listed above for version 0.36.

**Comment by caioproiete** [comment]
Thanks @govert. I'll get back to you on the topics below. Unfortunately it seems you were replying to my post at the same time I was editing it and adding more stuff, so there are some items in the different themes that I'm assuming you didn't see because you didn't comment on it. When you have a chance, could you see if you spot anything else you want to comment on?

**Comment by govert** [comment]
@caioproiete OK - I've updated my responses according to your edits.

I'm not sure how to structure the discussion - not ready for every item to be an individual issue though.
We'll see if anyone else chimes in over time.

**Comment by caioproiete** [comment]
@govert Great, thanks. Not sure how to structure the discussion either other than either creating individual issues and/or opening a channel on [Gitter](https://gitter.im) or [Slack](https://slack.com) to trash things out.

I'll wait a little to see if anyone else wants to give input - their own wish lists and/or comments to anything we've outlined above.

/cc @edparcell, @mndrake, @Ron-Ldn, @konne, @imanushin, @kchen0723, @Wildbiftek, @rlethbridge, @alainbryden, @sizzles, @phrrngtn, @autobotsassemble, @rosecodym, @masariello, @andrewkittredge, @dreamershl, @tomikiss, @hell-racer, @Mad-Lynx, @markns, @MisesEnForce, @WestXu, @vincentbeaulieu, @PGsHub, @ittegrat, @buybackoff, @akothoor, @davidandersen, @ian-cole-mcdavid

**Comment by phrrngtn** [comment]
I am happy to see this discussion and I am definitely interested in contributing some suggestions. The one which has been nagging me for years is a 'modern web browser in a CTP' with some thought given to making bidirectional communication between the browser and Excel tractable. 

I like the sound of a bunch of @caioproiete  suggestions but it will take me some time to go through them. I will also dust off some suggestions that I have sent directly to Govert over the years and see how they fit in (and if they are still relevant).

Excel DNA is an excellent product and I think that a great many useful things can be built around and on top of it.

**Comment by alainbryden** [comment]
Great discussion.  I may have some ideas to contribute in the near future (ideas to support installers, auto-updates, managing user settings, etc).

This one caught my eye and I felt the need to comment on it:

> Add support for (automatically) embedding metadata (and code) in Excel files, so that when a user sends a spreadsheet to someone that does not have the add-in installed, the user (at least) gets a nice message saying that they need to install add-in X for the spreadsheet to work.

We usually handle this by having a cell in a prominent place containing a formula like:
`=IFERROR(IF(OR(TRUE,ARe.SomeUDF()),"Thanks for using ARe"),"Load the add-in! analyzere.com")`

If you wanted to automatically do that to all your formulas, you could probably have a (pretty invasive) event handler that, On Save:
1. Bulk search/replaces all formulas containing UDFs from your add-in to wrap them with something like the above,
2. Then saves the workbook,
3. Then *undoes* the change so the user can be none the wiser.

You would then pair that with a workbook-open event handler that also does step 3.

**Comment by konne** [comment]
Hi, I really like this discussion but not the form. Be onest after was finished reading, I already lost the most of the requested features.
I would like to have more an issue related discussion. @govert @caioproiete would it be okay for you to create for each suggestion an issue. Please add Releases / Labels and just add them. (like Help Wanted,...)
An then just use features like thumb up/down,... do get an feedback of the users.

I agree to the most of your mentioned feature, and I think they just have to be ordered and we have to try to find contributors.

The nice think on my side is that I have now more R&D Power on my Excel Plugin so that I can even take one of my employes and let him develop some parts. The build process was in the last time mostly changed from caiproiete so that I didn't touch the msbuild task, but that is something that we can add very fast.

**Comment by govert** [comment]
@konne I agree - I'll wait a little while for any more feedback here, then split into a wiki page and separate issues.

Any particular features that you are interesting in seeing or working on?

**Comment by konne** [comment]
a real installer / helper for installing of the plugins
multi language for autocompletion / help context

**Comment by govert** [comment]
@konne Have you tried the WiXInstaller?

Can you explain the multi-language thing?

**Comment by caioproiete** [comment]
@phrrngtn: Love the "modern web browser in a CTP" idea. It would be really useful. In the past I ran into the [same issue that you did with CefSharp](https://github.com/cefsharp/CefSharp/issues/351) back in 2016, which to this day is still not resolved... Would be great to put together a team to either fix this and make it work, and/or explore other alternatives such as [headless Chrome](https://developers.google.com/web/updates/2017/04/headless-chrome) that didn't exist back then.

@alainbryden: I really like the simplicity of using an Excel formula to show the message about the add-in, and I've been using something along those lines coupled with worksheet properties that store the add-in name and version inside the worksheet, and inject some VBA that then gets used. Would like to explore this area more.

**Comment by konne** [comment]
> Have you tried the WiXInstaller?
yes, I already build some Intallers, I like to have a general Template with a cool WixInstaller hostest in /Excel-DNA/

>Can you explain the multi-language thing?
You can add for UDFs Names / Descriptions,.. this should be possible to have a out of the box multi language support.

**Comment by phrrngtn** [comment]
@andrewkittredge I was unable to get this going in my environment (I strongly suspect an application virtualization mechanism in use at the time).  I have to confess that I was not sure what needed to be done in order to get it up and running. 

@caioproiete I work in Reinsurance and will try and look you up the next time I am in Bermuda.

**Comment by caioproiete** [comment]
>
> @caioproiete I work in Reinsurance and will try and look you up the next time I am in Bermuda.
>

Small world! Yes, please do. I'll send you an e-mail with my contact info.

**Comment by caioproiete** [comment]
**Idea**: What about something like [FeatureUpvote](https://exceldna.featureupvote.com/) (similar to [UserVoice](https://www.uservoice.com) _except it's free for us\*_) to gauge votes from everyone here (we could send a link to the Google Groups too) to help us prioritize the discussions/items we want to scope out which ultimately can influence what goes in the roadmap and when?

### <https://exceldna.featureupvote.com>

[![image](https://user-images.githubusercontent.com/177608/40406446-153a12ba-5e37-11e8-8b25-2f56ae149460.png)](https://exceldna.featureupvote.com)

---

(\*) Steve McLeod, founder of [FeatureUpvote](http://featureupvote.com), was nice enough to give us a free plan after I signed up for a 30-day trial to experiment with it for Excel-DNA.

>
> From: Steve McLeod
> Date: Tue, May 22, 2018 at 8:09 PM
> Subject: Re: Feature Upvote for open-source project - Excel-DNA
> To: Caio Proiete
> 
> Hi Caio,
> 
> It is our pleasure to help Excel-DNA with tracking feature requests.
> 
> Your account is now on our "freebie" plan and ready to use.
> 
> Regards,
> 
> Steve McLeod
> Founder, Feature Upvote
>

**Comment by phrrngtn** [comment]
wrt browser in a CTP, I saw Webview component based on Edge: https://blogs.windows.com/msedgedev/2018/05/09/modern-webview-winforms-wpf-apps/

Apparently, this is part of Windows 10 1803.  It would be great if this 'just worked' with ExcelDNA.

**Comment by SeeSpotRun** [comment]
> >  Add support for developing add-ins with Visual Studio Code

> Not Sure / Sample / Independent Project - What does this mean?

I assume this means providing tools and templates (if required), and a tutorial / worked example, of how to develop and ExcelDNA Add-in using the [Visual Studio Code](https://code.visualstudio.com/) editor, instead of having to use the full [Visual Studio](https://visualstudio.microsoft.com/) IDE.

**Comment by ekanna** [comment]
FYI

Announcing .NET Core 3.0 on Sept 23, 2019
https://devblogs.microsoft.com/dotnet/announcing-net-core-3-0/

Next release after .NET Core 3.0 will be .NET 5
https://devblogs.microsoft.com/dotnet/introducing-net-5/

**Comment by Pinox** [comment]
I see there are lot of new com interop functionality in .net 5 - recently announced.
https://devblogs.microsoft.com/dotnet/improvements-in-native-code-interop-in-net-5-0/

**Comment by y2kbugger** [comment]
It looks like the dynamic COM (roadblock #1) has been resolved here: https://github.com/dotnet/runtime/pull/33060
